Improve jnp.split API

Sorry for the very uninformative title for this issue, couldn't think of a better one.

While reading the code for `split` in `jax/_src/numpy/lax_numpy.py`, I noticed the following block of code for the function `lax_numpy._split`:

https://github.com/google/jax/blob/2b716f292d814fdfadca43e108b5fedc122c0d2c/jax/_src/numpy/lax_numpy.py#L1178-L1193

Accounting to the type signature of `indices_or_sections`

https://github.com/google/jax/blob/2b716f292d814fdfadca43e108b5fedc122c0d2c/jax/_src/numpy/lax_numpy.py#L1172-L1173

We should be able to remove the first branch (i.e. `(tuple, list)`), in the spirit of #7737. Was about to send a pull request. But I also noticed that `indices_or_sections` is always concrete, and `Array`s are not hashable. Keeping the current type also prevents us from marking `indices_or_sections` as static in `jit`. One might argue removing the second branch(i.e. `(np.ndarray, ndarray)`) and make `indices_or_sections: int | Sequence[int]` made more sense, but that's a deviation from the original numpy api.

Not sure what's the best action to take here.

I ran into this issue. I want to pass a `list` of plain Python `int`s to a `jnp.split` operation inside a `jit`. (The list is constructed by a complex series of function calls.) If I pass the list directly, I get a mypy error:
```
error: Argument 2 to "split" has incompatible type "List[int]"; expected "Union[int, Union[Array, ndarray[Any, Any], bool_, number[Any], bool, int, float, complex]]"  [arg-type]
```
If I apply `jnp.array` to the list, I get a runtime error:
```
jax._src.errors.ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=6/0)>
in jax.numpy.split argument 1
```
What's the recommended solution to this? Is there a plan to change the type annotation of `jnp.split`?

--- a/jax/_src/numpy/lax_numpy.py
+++ b/jax/_src/numpy/lax_numpy.py
@@ -1154,7 +1154,8 @@ def broadcast_to(array: ArrayLike, shape: Shape) -> Array:
   return util._broadcast_to(array, shape)
 
 
-def _split(op: str, ary: ArrayLike, indices_or_sections: Union[int, ArrayLike],
+def _split(op: str, ary: ArrayLike,
+           indices_or_sections: Union[int, Sequence[int], ArrayLike],
            axis: int = 0) -> List[Array]:
   util.check_arraylike(op, ary)
   ary = asarray(ary)
@@ -1193,12 +1194,13 @@ def _split(op: str, ary: ArrayLike, indices_or_sections: Union[int, ArrayLike],
           for start, end in zip(split_indices[:-1], split_indices[1:])]
 
 @util._wraps(np.split, lax_description=_ARRAY_VIEW_DOC)
-def split(ary: ArrayLike, indices_or_sections: Union[int, ArrayLike], axis: int = 0) -> List[Array]:
+def split(ary: ArrayLike, indices_or_sections: Union[int, Sequence[int], ArrayLike],
+          axis: int = 0) -> List[Array]:
   return _split("split", ary, indices_or_sections, axis=axis)
 
 def _split_on_axis(op: str, axis: int) -> Callable[[ArrayLike, Union[int, ArrayLike]], List[Array]]:
   @util._wraps(getattr(np, op), update_doc=False)
-  def f(ary: ArrayLike, indices_or_sections: Union[int, ArrayLike]) -> List[Array]:
+  def f(ary: ArrayLike, indices_or_sections: Union[int, Sequence[int], ArrayLike]) -> List[Array]:
     # for 1-D array, hsplit becomes vsplit
     nonlocal axis
     util.check_arraylike(op, ary)
@@ -1213,7 +1215,8 @@ hsplit = _split_on_axis("hsplit", axis=1)
 dsplit = _split_on_axis("dsplit", axis=2)
 
 @util._wraps(np.array_split)
-def array_split(ary: ArrayLike, indices_or_sections: Union[int, ArrayLike], axis: int = 0) -> List[Array]:
+def array_split(ary: ArrayLike, indices_or_sections: Union[int, Sequence[int], ArrayLike],
+                axis: int = 0) -> List[Array]:
   return _split("array_split", ary, indices_or_sections, axis=axis)
 
 @util._wraps(np.clip, skip_params=['out'])

